src/api.jl

```diff
@@ -50,33 +50,70 @@ function Base.show(io::IO, e::HTTPError)
     end
 end
 
+"""
+    wait_until_done(ctx::Context, rsp::TransactionResponse)
+    wait_until_done(ctx::Context, transaction)
+    wait_until_done(ctx::Context, txn_id::String)
+
+Block until the provided `transaction` has reached a terminal state.
+
+Continuously polls get_transaction() for the transaction's state, until the transaction has
+finished. A transaction has finished once it has reached one of the terminal states:
+`COMPLETED` or `ABORTED`. The polling uses a low-overhead exponential backoff in order to
+ensure low-latency results without overloading network traffic.
+"""
+function wait_until_done(ctx::Context, rsp::TransactionResponse)
+    wait_until_done(ctx, rsp.transaction)
+end
+function wait_until_done(ctx::Context, txn::JSON3.Object)
+    wait_until_done(ctx, transaction_id(txn))
+end
+function wait_until_done(ctx::Context, id::AbstractString)
+    try
+        _poll_with_specified_overhead(overhead_rate = 0.01) do
+            txn = get_transaction(ctx, id)
+            return transaction_is_done(txn)
+        end
+        t = @spawn get_transaction(ctx, id)
+        m = @spawn get_transaction_metadata(ctx, id)
+        p = @spawn get_transaction_problems(ctx, id)
+        r = @spawn get_transaction_results(ctx, id)
+
+        return TransactionResponse(fetch(t), fetch(m), fetch(p), fetch(r))
+    catch
+        # Always print out the transaction id so that users can still get the txn ID even
+        # if there's an error during polling (such as an InterruptException).
+        @error "Client-side error while executing transaction." transaction_id=id
+        rethrow()
+    end
+end
+
 # Polls until the execution `f()` is truthy or the maximum number of polls is
-# reached. Polling frequency is controlled by an `ExponentialBackOff`. If `throw_on_max_n`
-# is set to true, this will throw if the maximum number of iterations are reached.
-function _poll_until(
+# reached. Polling frequency is controlled to minimize overhead, by keeping the overhead
+# from sleeping within a specified overhead rate. If `throw_on_max_n` is set to true, this
+# will throw if the maximum number of iterations are reached.
+function _poll_with_specified_overhead(
     f;
-    n=typemax(Int), # Maximum number of polls
-    first_delay=0.5,
-    factor=1.1,
-    max_delay=120, # 2 min
-    throw_on_max_n=false,
+    overhead_rate,  # Add xx% overhead through polling.
+    n = typemax(Int), # Maximum number of polls
+    max_delay = 120, # 2 min
+    throw_on_max_n = false,
 )
-    backoff = Base.ExponentialBackOff(
-        n=n,
-        first_delay=first_delay,
-        factor=factor,
-        max_delay=max_delay,
-    )
-
-    for duration in backoff
+    @assert overhead_rate >= 0.0
+    delay_rate = 1.0 + overhead_rate
+    start_time = time_ns()
```

Comment:
This should be from the moment the transaction request is received (is there a timestamp we can access)?

Reply:
Oh I guess this is client-side so it doesn't really matter?

Reply:
Yeah, exactly. The client is _creating_ the transaction, so that start time, _really is_ the start time. 👍

Reply:
ahhhh but no you're right, this was _not_ including the time since we first created the TXN in the calculation for how long to poll for.
You're right! Okay let me see if i can work that in.
If not, the `txn` object that was returned from the server actually _does_ include the start time from the server's perspective, so in the worst case we can use that. Good catch, @tveldhui! Thanks